### ISSUE TITLE
\React\Promise\all not preserving array key order

### ISSUE THREAD

**Issue by lordthorzonus** [reactphp/promise#74]
Hi I ran into an strange issue and am wondering is this intended behavior?

Basically I'm expecting that for the included code the output would be:

```
.Array
(
    [0] => 1
    [1] => 2
    [2] => 3
)
```

But instead I'm getting the array keys unordered:
```
.Array
(
    [0] => 1
    [2] => 3
    [1] => 2
)
```

Which causes problems for example when json_encoding (PHP thinks it should be encoded as an object not an array).

Code to reproduce:

```php
$loop = Factory::create();

$testData = [
    new Promise(
        function ($resolve) use ($loop) {
            $loop->nextTick(
                function () use ($resolve) {
                    $resolve(1);
                }
            );
        }
    ),
    new Promise(
        function ($resolve) use ($loop) {
            $loop->nextTick(
                function () use ($resolve, $loop) {
                    \React\Promise\resolve()->then( function() use ($resolve, $loop) {
                        $loop->nextTick(function () use ($resolve) {
                            $resolve(2);
                        });
                    });
                }
            );
        }
    ),
    new Promise(
        function ($resolve) use ($loop) {
            $loop->nextTick(
                function () use ($resolve) {
                    $resolve(3);
                }
            );
        }
    )
];


\React\Promise\all($testData)->then(function ($result) {
    print_r($result);
});

$loop->run();
``` 

Thanks!

**Comment by arosolino** [comment]
Seems the order of the values correlates to the order they are resolved. Couldn't you just sort the result?

```php
\React\Promise\all($testData)->then(function ($result) {
    ksort($result);
    print_r($result);
});
```

**Comment by vladar** [comment]
@arosolino this [explaination of ES6 spec](http://stackoverflow.com/questions/28066429/promise-all-order-of-resolved-values) is in favor of strictly preserving original ordering of elements. 

But even without ES6 spec... It seems useless to order final array after the order of promise resolution (as it is never guaranteed, which basically means that final array is returned sorted randomly). Is there any value in preserving the order of promise resolution vs original ordering?

**Comment by lordthorzonus** [comment]
@arosolino @vladar 

And the original order is also preserved simulating the same kind of test in chrome with JS:

```js
let p1 = Promise.resolve(1);
let p2 = Promise.resolve(2);
let p3 = new Promise(function (resolve, reject) {
    window.setTimeout(function () {
        Promise.resolve().then(function () {
                window.setTimeout(function () {
                    resolve(3)
                }, 250)
            }
        )
    }, 250);
});

Promise.all([p1, p2, p3]).then(values => {
    console.log(values); // [1, 2, 3]
});
```

**Comment by lordthorzonus** [comment]
@arosolino @vladar 

And the original order is also preserved simulating the same kind of test in chrome with JS:

```js
let p1 = new Promise(function (resolve, reject) {
    window.setTimeout(function () {
        resolve(1);
    }, 300);
});
let p2 = new Promise(function (resolve, reject) {
    window.setTimeout(function () {
        resolve(2);
    }, 100);
});
let p3 = new Promise(function (resolve, reject) {
    window.setTimeout(function () {
        Promise.resolve().then(function () {
                window.setTimeout(function () {
                    resolve(3)
                }, 250)
            }
        )
    }, 250);
});

Promise.all([p1, p2, p3]).then(values => {
    console.log(values); // [1, 2, 3]
});
```

**Comment by arosolino** [comment]
If the order ends up getting preserved an easy fix to this line could be:
https://github.com/reactphp/promise/blob/master/src/functions.php#L143

To:
```php
$values = array_fill(0, $toResolve, null);
```
Basically just prepopulate an array so order is created before anything resolves.

**Comment by arosolino** [comment]
If this needs to be done, then how about this.

https://github.com/reactphp/promise/blob/master/src/functions.php#L143

Instead of making a new $values array why not just overwrite each $promisesOrValues element with the resolved value. It would preserve order.

**Comment by jsor** [comment]
Yes, order is not preserved at the moment, only keys are preserved. 

> The order of the values correlates to the order they are resolved.

Yes, that is correct.

I'm not sure changing this would be a BC break, eg. for someone relying on that ordering. I'd like to hear opinions of @clue and @WyriHaximus on this.

**Comment by lordthorzonus** [comment]
@jsor I'm happy to provide a PR if this can be changed :). I think nobody can rely on the order being random though currently?

**Comment by jsor** [comment]
Yes, relying on the ordering would be very weird. But we're very conservative with BC. The question is not about the change itself but whether that should go into 2.x or 3.x.

**Comment by arosolino** [comment]
To preserve BC you could pass an extra parameter to all(), and then to map() for $preserveKeys.

**Comment by clue** [comment]
I agree that it makes sense to keep the original key order instead of the order the promises are resolved in :+1: Note that this can't really be compared with JavaScript, because they use `array` as an ordered list of values so that the key position is indeed significant, whereas PHP uses an hash-map with the original keys so that the key-value-relation is (more) important here.

> I'm not sure changing this would be a BC break, eg. for someone relying on that ordering. I'd like to hear opinions of @clue and @WyriHaximus on this.

I'd rather get this merged in the v2 branch, given that this behavior is currently undefined, undocumented and untested, I would not consider this a BC break :+1:

**Comment by WyriHaximus** [comment]
Agreeing with @clue here, this behavior is undocumented and thus changing to keep the array order isn't a BC break. Would also prefer to see this in the v2 branch (and v3 of course), as this is a neat little feature to have and  definitely have use cases for this :+1: .

**Comment by jsor** [comment]
@lordthorzonus Ok, appreciating a PR against the 2.x branch :)

**Comment by jsor** [comment]
Implemented by @lordthorzonus in #77

**Comment by jsor** [comment]
Released in [v2.5.0](https://github.com/reactphp/promise/releases/tag/v2.5.0).
